{{Infobox
 |name        = 貝葉氏譜系分析
 |bodystyle   = width:26em
 |titlestyle  = 
 |title       = 貝葉氏譜系分析
|headerstyle = 
 |labelstyle  = width:33%
 |datastyle   = 
 |header1 = 
 | label1 = 分類
 |  data1 = [[演化生物学|演化生物学]]
 |header2 = 
 | label2 = 子分類
 |  data2 = [[分子系統發生學|分子系統發生學]]
 |header3 = 
 | label3 = 最佳搜索條件
 |  data3 = [[贝叶斯推断|贝叶斯推断]]
}}
'''貝葉氏譜系分析'''（Bayesian inference of phylogeny）是根據已知的[[先驗機率|先驗機率]]推導[[後驗機率|後驗機率]]的[[似然函數|似然函數]]，進而去預測最可能的系譜樹。隨著電腦計算速度的演進以及[[马尔科夫蒙特卡洛|马尔科夫-蒙特卡洛法]]的演進，貝葉氏譜系分析越來越受重視和應用。[[贝叶斯推断|贝叶斯推断]]時常應用於[[分子系統發生學|分子系統發生學]]以及[[系統分類學|系統分類學]]。
==分析軟體列表==
下表列出最常用的系譜分析軟體，有些不只使用貝葉氏譜系分析。

{| class="wikitable"
|-
!軟體!! 描述 !! 方法 !! 作者 !! 網址
|-
|Armadillo Workflow Platform || Workflow platform dedicated to phylogenetic and general bioinformatic analysis || Inference of phylogenetic trees using Distance, Maximum Likelihood, Maximum Parsimony, Bayesian methods and related workflows || E. Lord, M. Leclercq, A. Boc, A.B. Diallo and V. Makarenkov <ref>Lord, E., Leclercq, M., Boc, A., Diallo, A.B., Makarenkov, V. 2012. Armadillo 1.1: An Original Workflow Platform for Designing and Conducting Phylogenetic Analysis and Simulations. PLoS ONE 7(1): e29903. doi:10.1371/journal.pone.0029903</ref> || http://www.bioinfo.uqam.ca/armadillo.
|-
|Bali-Phy || Simultaneous Bayesian inference of alignment and phylogeny  || Bayesian inference, alignment as well as tree search || M.A. Suchard, B. D. Redelings <ref>Suchard, M.A. and Redelings, B.D. 2006. BAli-Phy: simultaneous Bayesian inference of alignment and phylogeny, Bioinformatics. 22:2047-2048</ref> || http://www.bali-phy.org
|-
|BATWING || Bayesian Analysis of Trees With Internal Node Generation || Bayesian inference, demographic history, population splits || I. J. Wilson, D. Weale, D.Balding <ref>Wilson, I., Weale, D. and Balding, M. 2003. Inferences from DNA data: population histories, evolutionary processes and forensic match probabilities. Journal of the Royal Statistical Society: Series A (Statistics in Society), 166: 155-188</ref> || http://www.maths.abdn.ac.uk/˜ijw{{dead link|date=2018年4月 |bot=InternetArchiveBot |fix-attempted=yes }}
|-
| Bayes Phylogenies || Bayesian inference of trees using Markov Chain Monte Carlo methods || Bayesian inference, multiple models, mixture model (auto-partitioning) || M. Pagel, A. Meade <ref> Pagel, M. and Meade, A. 2006. Bayesian analysis of correlated evolution of discrete characters by reversible-jump Markov chain Monte Carlo. American Naturalist, 167, 808-825</ref> || http://www.evolution.rdg.ac.uk/BayesPhy.html
|-
|PhyloBayes / PhyloBayes MPI || Bayesian Monte Carlo Markov Chain (MCMC) sampler for phylogenetic reconstruction. || Non-parametric methods for modeling among-site variation in nucleotide or amino-acid propensities. || N. Lartillot, N. Rodrigue, D. Stubbs, J. Richer <ref> Lartillot N., Philippe H. 2004. A Bayesian Mixture Model for Across-Site Heterogeneities in the Amino-Acid Replacement Process. Molecular Biology and Evolution 21(6):1095-1109</ref> || http://www.phylobayes.org
|-
| BEAST || Bayesian Evolutionary Analysis Sampling Trees || Bayesian inference, relaxed molecular clock, demographic history || A. J. Drummond, A. Rambaut & M. A. Suchard <ref>Drummond, A.J., Rambaut, A. 2007. Beast: Bayesian evolutionary analysis by sampling trees. BMC Evol Biol. 7:214</ref> || http://beast.bio.ed.ac.uk
|-
| BUCKy || Bayesian concordance of gene trees || Bayesian concordance using modified greedy consensus of unrooted quartets || C. Ané, B. Larget, D.A. Baum, S.D. Smith, A. Rokas and B. Larget, S.K. Kotha, C.N. Dewey, C. Ané <ref>Ané, C., Larget, B., Baum, D.A.,Smith, S.D., Rokas, A. 2007. Bayesian estimation of concordance among gene trees. Molecular Biology and Evolution 24(2), 412-426</ref> || http://www.stat.wisc.edu/~ane/bucky/
|-
|Geneious (MrBayes plugin) || Geneious provides genome and proteome research tools || Neighbor-joining, UPGMA, MrBayes plugin, PHYML plugin, RAxML plugin, FastTree plugin, GARLi plugin, PAUP* Plugin || A. J. Drummond,M.Suchard,V.Lefort et al. || http://www.geneious.com
|-
| TOPALi || Phylogenetic inference || Phylogenetic model selection, Bayesian analysis and Maximum Likelihood phylogenetic tree estimation, detection of sites under positive selection, and recombination breakpoint location analysis || I.Milne, D.Lindner, et al.<ref>Milne, I., Lindner, D., Bayer, M., Husmeier, D., McGuire, G., Marshall, D.F. and Wright, F. 2008. TOPALi v2: a rich graphical interface for evolutionary analyses of multiple alignments on HPC clusters and multi-core desktops. Bioinformatics 25 (1), 126-127</ref> || http://www.topali.org

|}

== 應用 ==
貝葉氏譜系分析在分子系統發生學上具有相當廣泛的應用，包含：
[[File:Divergence_time_estimation_and_ancestral_area_reconstruction_of_porcini_s.s..png|thumb]]（molecular clock analysis）得出的系譜樹。圓餅圖表示的是利用貝葉氏二元MCMC分析（BBM）所推斷的祖源。]]

* 系統發生學預測<ref>Alonso, R., Crawford, A.J. & Bermingham, E. 2011. Molecular phylogeny of an endemic radiation of
Cuban toads (Bufonidae: Peltophryne) based on mitochondrial and nuclear genes. Journal of
Biogeography 39 (3): 434-451
</ref><ref>Antonelli, A., Sanmart.n, I. 2011. Mass Extinction, Gradual Cooling, or Rapid Radiation? Reconstructing the Spatiotemporal Evolution of the Ancient Angiosperm Genus Hedyosmum (Chloranthaceae) Using Empirical and Simulated Approaches. Syst. Biol. 60(5):596–615</ref>
* 預測及評估未明祖源<ref>de Villemereuil, P.,Wells, J.A., Edwards, R.D. and Blomberg, S.P. 2012. Bayesian Phylogeography Finds Its Roots BMC Evolutionary Biology 12:102</ref>
*祖徵演化<ref>Ronquist, F. 2004. Bayesian inference of character Evolution. Trends in Ecology and Evolution 19 No.9: 475-481</ref><ref>Schäffer , S., Koblmüller, S., Pfingstl, T., Sturmbauer, C., Krisper, G. 2010. Ancestral state reconstruction reveals multiple independent evolution of diagnostic morphological characters in the “Higher Oribatida” (Acari), conflicting with current classification schemes. BMC Evolutionary Biology 10:246</ref>
* 祖源地域推論<ref>Filipowicz, N., Renner, S. 2012. Brunfelsia (Solanaceae): A genus evenly divided between South America and radiations on Cuba and other Antillean islands. Molecular Phylogenetics and Evolution 64: 1-11</ref>
* 分子測年分析<ref>Bacon, C.D., Baker, W.J., Simmons, M.P. 2012a. Miocene dispersal drives island radiations in the
palm tribe Trachycarpeae (Arecaceae). Systematic Biology 61: 426–442</ref><ref>Särkinen, T., Bohs, L., Olmstead,R.G. and Knapp, S. 2013. A phylogenetic framework for evolutionary study of the nightshades (Solanaceae): a dated 1000-tip tree. BMC Evolutionary Biology. 13:214</ref>
* 物種多樣化和滅絕的動力學模型<ref>Silvestro, D., Schnitzler, J., Liow, L.H., Antonelli, A., Salamin, N.    2014. Bayesian Estimation of Speciation and Extinction from Incomplete Fossil Occurrence Data. Syst. Biol. 63(3):349–367</ref>
* 傳染病擴散模型<ref>Lemey, P., Rambaut, A., Drummond, A.J., Suchard, M.A. 2009. Bayesian Phylogeography Finds Its Roots. PLoS Comput Biol 5(9): e1000520</ref>

== 參考文獻 ==
{{Reflist}}

== 外部連結==
* [http://mrbayes.sourceforge.net/ MrBayes official website]
* [http://beast.bio.ed.ac.uk/ BEAST official website]

{{Phylogenetics}}

[[Category:生物信息學|Category:生物信息學]]
[[Category:计算系统发生学|Category:计算系统发生学]]
{{tsl|en|Category:Applications of Bayesian inference|Category:貝葉氏導出式分析相關應用}}