{{NoteAT|G1=生命科学}}
{{enzyme
| Name = 组蛋白-赖氨酸<br />''N''-甲基转移酶
| EC_number = 2.1.1.43
| CAS_number = 9055-08-7
| IUBMB_EC_number = 2/1/1/43
| GO_code = 0042054
| image = 
| width = 
| caption =
}}
'''组蛋白甲基转移酶'''（{{lang-en|Histone methyltransferase}}，简称为'''HMT'''）是包括[[EZH2|组蛋白-赖氨酸''N''-甲基转移酶]]与组蛋白-精氨酸''N''-甲基转移酶在内的是一大類组蛋白修饰[[酶|酶]]类，它们[[催化|催化]]将一个、两个或三个[[甲基|甲基]]转移到[[组蛋白|组蛋白]]的[[赖氨酸|赖氨酸]]或[[精氨酸|精氨酸]]残基上。被添加上的甲基基团主要位于组蛋白H3和H4的特定赖氨酸或精氨酸上<ref name=Posttranslational>{{cite book | editor = Conaway JW, Conaway RC | author = Wood A | title = Proteins in eukaryotic transcription | series = Advances in Protein Chemistry | volume = 67 | publisher = Elsevier Academic Press | location = Amsterdam | year = 2004 | chapter = Posttranslational Modifications of Histones by Methylation | pages =  201–222 | isbn = 0-12-034267-7 | doi = 10.1016/S0065-3233(04)67008-2 }}</ref>。

== 种类 ==
赖氨酸特异性组蛋白甲基转移酶类可被细分为含SET结构域的和不含SET结构域的两类。顾名思义，它们之间的区别在于存不存在SET这样一种结构域。

人类基因所编码的具有组蛋白甲基转移酶活性的蛋白质包括：
* [[ASH1L|ASH1L]]
* [[DOT1L|DOT1L]]
* [[EHMT1|EHMT1]]、[[EHMT2|EHMT2]]、[[EZH1|EZH1]]、[[EZH2|EZH2]]
* [[MLL|MLL]]、[[MLL2|MLL2]]、[[MLL3|MLL3]]、[[MLL4|MLL4]]、[[MLL5|MLL5]]
* [[NSD1|NSD1]]
* [[PRDM2|PRDM2]]
* [[SET_(基因)|SET]]、[[SETBP1|SETBP1]]、[[SETD1A|SETD1A]]、[[SETD1B|SETD1B]]、[[SETD2|SETD2]]、[[SETD3|SETD3]]、[[SETD4|SETD4]]、[[SETD5|SETD5]]、[[SETD6|SETD6]]、[[SETD7|SETD7]]、[[SETD8|SETD8]]、[[SETD9|SETD9]]、[[SETDB1|SETDB1]]、[[SETDB2|SETDB2]]
* [[SETMAR|SETMAR]]、[[SMYD1|SMYD1]]、[[SMYD2|SMYD2]]、[[SMYD3|SMYD3]]、[[SMYD4|SMYD4]]、[[SMYD5|SMYD5]]、[[SUV39H1|SUV39H1]]、[[SUV39H2|SUV39H2]]、[[SUV420H1|SUV420H1]]、[[SUV420H2|SUV420H2]]

=== 含SET结构域的赖氨酸特异性组蛋白甲基转移酶 ===

==== 结构 ====
影响甲基转移酶活性的结构包括：SET结构域（包括130个氨基酸）、前SET和后SET结构域。前SET区域包括[[半胱氨酸|半胱氨酸]]残基，可形成三角锌簇，紧密结合锌原子并使结构稳定。SET结构域本身包含一个富含β-股的催化核心，从而形成几个[[β折叠|β折叠]]区域。 一般在前SET结构域发现的β-股将形成的β-折叠，它带有SET结构域的β-股，导致SET结构域结构的细微变化。

These small changes alter the target residue site specificity for methylation and allow the SET domain methyltransferases to target many different residues. This interplay between the pre-SET domain and the catalytic core is critical for enzyme function.<ref name=Posttranslational />

==== 催化机制 ====

In order for the reaction to proceed, [[S-Adenosyl_methionine|S-Adenosyl methionine]] (SAM) and the lysine residue of the substrate histone tail must first be bound and properly oriented in the catalytic pocket of the SET domain.  Next, a nearby tyrosine residue deprotonates the ε-amino group of the lysine residue.<ref name=Trievel>{{cite journal | author = Trievel RC, Beach BM, Dirk LM, Houtz RL, Hurley JH | title = Structure and catalytic mechanism of a SET domain protein methyltransferase | journal = Cell | volume = 111 | issue = 1 | pages = 91–103 |date=October 2002 | pmid = 12372303 | doi = 10.1016/S0092-8674(02)01000-0| url = }}</ref>  The lysine chain then makes a nucleophilic attack on the methyl group on the sulfur atom of the SAM molecule, transferring the methyl group to the lysine side chain.

[[File:Histone_Methyltransferase,_SAM_and_Lysine_residue.tiff|thumb]]

=== 不含SET结构域的赖氨酸特异性组蛋白甲基转移酶 ===

Instead of SET, non-SET domain-containing histone methyltransferase utilizes the enzyme Dot1.  Unlike the SET domain, which targets the lysine tail region of the histone, Dot1 methylates a lysine residue in the globular core of the histone, and is the only enzyme known to do so.<ref name=Posttranslational /> A possible homolog of Dot1 was found in archaea which shows the ability to methylate archaeal histone-like protein in recent studies.

==== 结构 ====

The N terminal of Dot1 contains the active site.   A loop serving as the binding site for SAM links the N-terminal and the C-terminal domains of the Dot1 catalytic domain.  The C-terminal is important for the substrate specificity and binding of Dot1 because the region carries a positive charge, allowing for a favorable interaction with the negatively charged backbone of DNA.<ref name=Min>{{cite journal | author = Min J, Feng Q, Li Z, Zhang Y, Xu RM | title = Structure of the catalytic domain of human DOT1L, a non-SET domain nucleosomal histone methyltransferase | journal = Cell | volume = 112 | issue = 5 | pages = 711–23 |date=March 2003 | pmid = 12628190 | doi =10.1016/S0092-8674(03)00114-4  }}</ref>  Due to structural constraints, Dot1 is only able to methylate histone H3.

=== 精氨酸特异性组蛋白甲基转移酶 ===

There are two different types of protein arginine methyltransferases (PRMTs) and three types of methylation that can occur at arginine residues on histone tails.  The first type of PRMTs (PRMT1, PRMT3, CARM1⧸PRMT4, and Rmt1⧸Hmt1) produce monomethylarginine and asymmetric dimethylarginine.<ref name=Chen>{{cite journal | author = Chen D, Ma H, Hong H, Koh SS, Huang SM, Schurter BT, Aswad DW, Stallcup MR | title = Regulation of transcription by a protein methyltransferase | journal = Science | volume = 284 | issue = 5423 | pages = 2174–7 |date=June 1999 | pmid = 10381882 | doi =10.1126/science.284.5423.2174 }}</ref><ref name=Gary>{{cite journal | author = Gary JD, Lin WJ, Yang MC, Herschman HR, Clarke S | title = The predominant protein-arginine methyltransferase from Saccharomyces cerevisiae | journal = J. Biol. Chem. | volume = 271 | issue = 21 | pages = 12585–94 |date=May 1996 | pmid = 8647869 | doi =10.1074/jbc.271.21.12585  }}</ref><ref name=McBride>{{cite journal | author = McBride AE, Weiss VH, Kim HK, Hogle JM, Silver PA | title = Analysis of the yeast arginine methyltransferase Hmt1p/Rmt1p and its in vivo function. Cofactor binding and substrate interactions | journal = J. Biol. Chem. | volume = 275 | issue = 5 | pages = 3128–36 |date=February 2000 | pmid = 10652296 | doi =10.1074/jbc.275.5.3128 }}</ref>  The second type (JBP1⧸PRMT5) produces monomethyl or symmetric dimethylarginine.<ref name=Branscombe />  The differences in the two types of PRMTs arise from restrictions in the arginine binding pocket.<ref name=Branscombe /> 

==== 结构 ====

The catalytic domain of PRMTs consists of a SAM binding domain and substrate binding domain (about 310 amino acids in total).<ref name=Branscombe /><ref name=Weiss /><ref name=Zhang />  Each PRMT has a unique N-terminal region and a catalytic core.  The arginine residue and SAM must be correctly oriented within the binding pocket.  SAM is secured inside the pocket by a hydrophobic interaction between an adenine ring and a phenyl ring of a phenylalanine.<ref name=Zhang />  

==== 催化机制 ====

A glutamate on a nearby loop interacts with nitrogens on the target arginine residue. This interaction redistributes the positive charge and leads to the deprotonation of one nitrogen group,<ref name=McBride2>{{cite journal | author = McBride AE, Silver PA | title = State of the arg: protein methylation at arginine comes of age | journal = Cell | volume = 106 | issue = 1 | pages = 5–8 |date=July 2001 | pmid = 11461695 | doi = 10.1016/S0092-8674(01)00423-8 }}</ref> which can then make a nucleophilic attack on the methyl group of SAM.  Differences between the two types of PRMTs determine the next methylation step: either catalyzing the dimethylation of one nitrogen or allowing the symmetric methylation of both groups.<ref name=Branscombe />  However, in both cases the proton stripped from the nitrogen is dispersed through a histidine–aspartate proton relay system and released into the surrounding matrix.<ref name=Fersht>{{cite journal | author = Fersht AR, Sperling J | title = The charge relay system in chymotrypsin and chymotrypsinogen | journal = J. Mol. Biol. | volume = 74 | issue = 2 | pages = 137–49 |date=February 1973 | pmid = 4689953 | doi =10.1016/0022-2836(73)90103-4  }}</ref>

== 在基因调控中的作用 ==
[[Histone_methylation|Histone methylation]] plays an important role in [[epigenetic|epigenetic]] [[gene_regulation|gene regulation]].  Methylated histones can either repress or activate transcription as different experimental findings suggest. For example, it is likely that the methylation of lysine 9 on histone H3 (H3K9me3) in the promoter region of genes prevents excessive expression of these genes and, therefore, delays cell cycle transition and/or proliferation.<ref name=Methylation />  See [[Histone#Chromatin_regulation|Histone#Chromatin regulation]].

== 疾病的相关性 ==
Abnormal expression or activity of methylation-regulating enzymes has been noted in some types of human cancers, suggesting associations between [[histone_methylation|histone methylation]] and malignant transformation of cells or formation of tumors.<ref name=Methylation>{{cite book | author = Chen F, Kan H, Castranova V | editor = Tollefsbol TO | others = | title = Handbook of Epigenetics: The New Molecular and Medical Genetics | publisher = Academic Press | location = Boston | year = 2010| pages = 149–157 | chapter = Methylation of Lysine 9 of Histone H3: Role of Heterochromatin Modulation and Tumorigenesis | isbn = 0-12-375709-6  | doi = 10.1016/B978-0-12-375709-8.00010-1 }}</ref>   In recent years, epigenetic modification of the histone proteins, especially the methylation of the histone H3, in cancer development has been an area of emerging research.  It is now generally accepted that in addition to genetic aberrations, cancer can be initiated by epigenetic changes in which gene expression is altered without genomic abnormalities.  These epigenetic changes include loss or gain of methylations in both DNA and histone proteins.<ref name=Methylation />

There is not yet compelling evidence that suggests cancers develop purely by abnormalities in [[histone_methylation|histone methylation]] or its signaling pathways, however they may be a contributing factor. For example, down-regulation of methylation of lysine 9 on histone 3 (H3K9me3) has been observed in several types of human cancer (such as colorectal cancer, ovarian cancer, and lung cancer), which arise from either the deficiency of H3K9 methyltransferases or elevated activity or expression of H3K9 demethylases.<ref name=Methylation /><ref name=Espino>{{cite journal | author = Espino PS, Drobic B, Dunn KL, Davie JR | title = Histone modifications as a platform for cancer therapy | journal = J. Cell. Biochem. | volume = 94 | issue = 6 | pages = 1088–102 |date=April 2005 | pmid = 15723344 | doi = 10.1002/jcb.20387  }}</ref><ref name=Hamamoto>{{cite journal | author = Hamamoto R, Furukawa Y, Morita M, Iimura Y, Silva FP, Li M, Yagyu R, Nakamura Y | title = SMYD3 encodes a histone methyltransferase involved in the proliferation of cancer cells | journal = Nat. Cell Biol. | volume = 6 | issue = 8 | pages = 731–40 |date=August 2004 | pmid = 15235609 | doi = 10.1038/ncb1151 }}</ref>

== 进一步研究 ==
组蛋白甲基转移酶可以用作癌症的诊断和预后的生物标志物。 另外，关于组蛋白甲基转移酶在细胞的恶性转化，组织的癌发生和肿瘤发生中的功能和调节仍然存在许多问题<ref name=Methylation />。

== 另见 ==
*{{le|组蛋白修饰酶类|Histone-modifying enzymes}}
*[[组蛋白乙酰转移酶|组蛋白乙酰转移酶（HAT）]]
*[[组蛋白脱乙酰酶|组蛋白脱乙酰酶（HDAC）]]
*[[组蛋白甲基化|组蛋白甲基化]]
*[[核小體|核小體]]
*[[染色质|染色质]]

== 参考文献 ==
{{reflist|colwidth = 35em}}

== 深入阅读 ==
{{refbegin|colwidth=35em}}
* {{cite journal | author = Trievel RC | title = Structure and function of histone methyltransferases | journal = Crit. Rev. Eukaryot. Gene Expr. | volume = 14 | issue = 3 | pages = 147–69 | year = 2004 | pmid = 15248813 | doi = 10.1615/CritRevEukaryotGeneExpr.v14.i3.10| url = }}
* {{cite journal | author = Conde F, Refolio E, Cordón-Preciado V, Cortés-Ledesma F, Aragón L, Aguilera A, San-Segundo PA | title = The Dot1 histone methyltransferase and the Rad9 checkpoint adaptor contribute to cohesin-dependent double-strand break repair by sister chromatid recombination in Saccharomyces cerevisiae | journal = Genetics | volume = 182 | issue = 2 | pages = 437–46 |date=June 2009 | pmid = 19332880 | pmc = 2691753 | doi = 10.1534/genetics.109.101899  }}
{{refend}}

== 外部链接 ==
* [http://www.ncbi.nlm.nih.gov/bookshelf/br.fcgi?book=gene&part=kleefstra  GeneReviews/NCBI/NIH/UW entry on Kleefstra Syndrome]
* {{MeshName|Histone-Lysine+N-Methyltransferase}}
* {{MeshName|Protein-Arginine+N-Methyltransferase}}

{{-}}
{{EC 2.1}}
{{转录}}

[[Category:EC_2.1.1|Category:EC 2.1.1]]
[[Category:表觀遺傳學|Category:表觀遺傳學]]